### PR TITLE
Added WarmValency to local root configuration

### PR DESCRIPTION
**Pull request by bolt12**
- Refactor Governor established `belowTargetLocal` job
- Refactor progress to established target test

Solution for #4565

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs [hunk 1]
```diff
@@ -44,9 +50,23 @@ data LocalRootPeers peeraddr =
        (Map peeraddr PeerAdvertise)
 
        -- The groups, but without the associated PeerAdvertise
-       [(Int, Set peeraddr)]
+       [((HotValency, WarmValency), Set peeraddr)]
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs:53]: There's no need for extra indirection:
```suggestion
       [(HotValency, WarmValency, Set peeraddr)]
```

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 2]
```diff
@@ -70,16 +72,20 @@ import           Ouroboros.Network.PeerSelection.RootPeersDNS.DNSActions
...
 data TraceLocalRootPeers peerAddr exception =
-       TraceLocalRootDomains [(Int, Map RelayAccessPoint PeerAdvertise)]
+       TraceLocalRootDomains [( (HotValency, WarmValency)
+                              , Map RelayAccessPoint PeerAdvertise)]
        -- ^ 'Int' is the configured valency for the local producer groups
      | TraceLocalRootWaiting DomainAccessPoint DiffTime
      | TraceLocalRootResult  DomainAccessPoint [(IP, DNS.TTL)]
-     | TraceLocalRootGroups  [(Int, Map peerAddr PeerAdvertise)]
+     | TraceLocalRootGroups  [( (HotValency, WarmValency)
+                              , Map peerAddr PeerAdvertise)]
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:81]:
```suggestion
     | TraceLocalRootGroups  [(HotValency, WarmValency, Map peerAddr PeerAdvertise)]
```

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 3]
```diff
@@ -70,16 +72,20 @@ import           Ouroboros.Network.PeerSelection.RootPeersDNS.DNSActions
...
        -- ^ This traces the results of the local root peer provider
      | TraceLocalRootDNSMap  (Map DomainAccessPoint [peerAddr])
        -- ^ This traces the results of the domain name resolution
-     | TraceLocalRootReconfigured [(Int, Map RelayAccessPoint PeerAdvertise)] -- ^ Old value
-                                  [(Int, Map RelayAccessPoint PeerAdvertise)] -- ^ New value
+     | TraceLocalRootReconfigured [( (HotValency, WarmValency)
+                                   , Map RelayAccessPoint PeerAdvertise)] -- ^ Old value
+                                  [( (HotValency, WarmValency)
+                                   , Map RelayAccessPoint PeerAdvertise)] -- ^ New value
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:88]:
```suggestion
     | TraceLocalRootReconfigured [(HotValency, WarmValency, Map RelayAccessPoint PeerAdvertise)] -- ^ Old value
                                  [(HotValency, WarmValency, Map RelayAccessPoint PeerAdvertise)] -- ^ New value
```

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 4]
```diff
@@ -121,7 +129,8 @@ localRootPeersProvider tracer
     -- | Loop function that monitors DNS Domain resolution threads and restarts
     -- if either these threads fail or detects the local configuration changed.
     --
-    loop :: [(Int, Map RelayAccessPoint PeerAdvertise)] -> m Void
+    loop :: [((HotValency, WarmValency), Map RelayAccessPoint PeerAdvertise)]
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:132]:
```suggestion
    loop :: [(HotValency, WarmValency, Map RelayAccessPoint PeerAdvertise)]
```

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 5]
```diff
@@ -102,9 +108,11 @@ localRootPeersProvider
   -> (IP -> Socket.PortNumber -> peerAddr)
   -> DNS.ResolvConf
   -> DNSActions resolver exception m
-  -> STM m [(Int, Map RelayAccessPoint PeerAdvertise)]
+  -> STM m [( (HotValency, WarmValency)
+            , Map RelayAccessPoint PeerAdvertise)]
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:112]:
```suggestion
  -> STM m [(HotValency, WarmValency, Map RelayAccessPoint PeerAdvertise)]
```

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 6]
```diff
@@ -102,9 +108,11 @@ localRootPeersProvider
...
   -- ^ input
-  -> StrictTVar m [(Int, Map peerAddr PeerAdvertise)]
+  -> StrictTVar m [( (HotValency, WarmValency)
+                   , Map peerAddr PeerAdvertise)]
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:115]:
```suggestion
  -> StrictTVar m [(HotValency, WarmValency, Map peerAddr PeerAdvertise)]
```

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Simple.hs [hunk 7]
```diff
@@ -57,7 +59,8 @@ withPeerSelectionActions
   -> (IP -> Socket.PortNumber -> peeraddr)
   -> DNSActions resolver exception m
   -> STM m PeerSelectionTargets
-  -> STM m [(Int, Map RelayAccessPoint PeerAdvertise)]
+  -> STM m [( (HotValency, WarmValency)
+            , Map RelayAccessPoint PeerAdvertise)]
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Simple.hs:63]:
```suggestion
  -> STM m [(HotValency, WarmValency, Map RelayAccessPoint PeerAdvertise)]
```

File: ouroboros-network/test/Test/Ouroboros/Network/Diffusion/Node.hs [hunk 8]
```diff
@@ -147,7 +149,8 @@ data Arguments m = Arguments
     , aChainSyncEarlyExit   :: Bool
 
     , aPeerSelectionTargets :: PeerSelectionTargets
-    , aReadLocalRootPeers   :: STM m [(Int, Map RelayAccessPoint PeerAdvertise)]
+    , aReadLocalRootPeers   :: STM m [( (HotValency, WarmValency)
+                                      , Map RelayAccessPoint PeerAdvertise)]
```
Comment (coot) [inline ouroboros-network/test/Test/Ouroboros/Network/Diffusion/Node.hs:153]:
```suggestion
    , aReadLocalRootPeers   :: STM m [(HotValency, WarmValency, Map RelayAccessPoint PeerAdvertise)]
```

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 9]
```diff
@@ -220,7 +222,8 @@ data ArgumentsExtra m = ArgumentsExtra {
       --
       daPeerSelectionTargets :: PeerSelectionTargets
 
-    , daReadLocalRootPeers  :: STM m [(Int, Map RelayAccessPoint PeerAdvertise)]
+    , daReadLocalRootPeers  :: STM m [( (HotValency, WarmValency)
+                                      , Map RelayAccessPoint PeerAdvertise)]
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:226]:
```suggestion
    , daReadLocalRootPeers  :: STM m [(HotValency, WarmValency, Map RelayAccessPoint PeerAdvertise)]
```

**Review by coot — CHANGES_REQUESTED** [review]

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs:53]: It is going to be set as the same number as whatever was specified for `HotValency`, so by default `WarmValency = HotValency`

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs [hunk 10]
```diff
@@ -102,30 +131,36 @@ toGroupSets (LocalRootPeers _ gs) = gs
...
+    establishStructureInvariant !acc (((h, w), g): gs)
+      | w' > 0 && h' > 0 = ((h', w'), g') : establishStructureInvariant acc' gs
+      | otherwise        =                  establishStructureInvariant acc' gs
       where
         !g'   = g `Map.withoutKeys` acc
-        !t'   = min t (Map.size g')
+        !w'   = min w (WarmValency (Map.size g'))
+        !h'   = if getHotValency h > getWarmValency w'
+                   then coerce w'
+                   else h
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs:155]: Yeah that's cleaner :)

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs [hunk 11]
```diff
@@ -690,7 +692,7 @@ data TracePeerSelection peeraddr =
      -- | target active, actual active, selected peers
      | TraceDemoteHotPeers     Int Int (Set peeraddr)
      -- | local per-group (target active, actual active), selected peers
-     | TraceDemoteLocalHotPeers [(Int, Int)] (Set peeraddr)
+     | TraceDemoteLocalHotPeers [(HotValency, Int)] (Set peeraddr)
```
[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:695]: Okay, I only refactored this to use the newtype instead, I'll rebase on top of your patch once it is merged

[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:696]: Don't wait, it's in `coot/eclipse-evasion`. I'll resolve conflicts.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs]: Having some slack is actually good. If I recall now, the solution we agreed on was `WarmValency + 1` but it should be ```(WarmValency + 1) `min` length```.

[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs]: The default I recall agreeing was `WarmValency = HotValency` and then strongly suggest users to configure it as `WarmValency = HotValency + 1` if they want some fallback nodes. But I am also in favour of making it as you suggested!

[the remark this answers is not in the document]
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs]: Anyway this doesn't affect this PR only the cardano-nos counterpart

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs [hunk 12]
```diff
@@ -74,49 +76,48 @@ belowTargetLocal actions
...
-    -- in the connect set. We can also subtract the in progress ones since they
-    -- are also already in the connect set and we cannot pick them again.
-  , numLocalAvailableToConnect - numLocalEstablishedPeers
-                               - numLocalConnectInProgress > 0
-  --TODO: switch style to checking if the set is empty
+    -- Are there any groups of local peers that are below target?
+  | not (null groupsBelowTarget)
+    -- We need this detailed check because it is not enough to check we are
+    -- below an aggregate target. We can be above target for some groups
+    -- and below for others.
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs:87]: >  We can be above target for some groups and below for others.

I think we need this other check for another reason. We need to take into account peers which are being promoted to `Warm`, and peers which are being demoted to `Cold`.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs [hunk 13]
```diff
@@ -74,49 +76,48 @@ belowTargetLocal actions
...
+
+    -- Are there any groups where we can pick members to promote?
+  , let groupsAvailableToPromote =
+          [ (numMembersToPromote, membersAvailableToPromote)
+          | let availableToPromote =
+                  localAvailableToConnect
+                     Set.\\ localEstablishedPeers
+                     Set.\\ localConnectInProgress
+                     Set.\\ localDemoteInProgress
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs:96]: Should we take into account `localDemoteInProgress`?  This means we can go above the target while we are demoting some peers, which I think we should avoid. Note that below in `numMembersToPromote` we don't take demotions to `Cold` into account.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs [hunk 14]
```diff
@@ -44,9 +49,23 @@ data LocalRootPeers peeraddr =
...
+newtype HotValency = HotValency { getHotValency :: Int }
+  deriving (Show, Eq, Ord)
+  deriving Num via Int
+
+-- | Newtype wrapper representing warm valency value from local root group
+-- configuration
+--
+newtype WarmValency = WarmValency { getWarmValency :: Int }
+  deriving (Show, Eq, Ord)
+  deriving Num via Int
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/LocalRootPeers.hs:67]: Nice.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 15]
```diff
@@ -70,16 +72,24 @@ import           Ouroboros.Network.PeerSelection.RootPeersDNS.DNSActions
...
      | TraceLocalRootDNSMap  (Map DomainAccessPoint [peerAddr])
        -- ^ This traces the results of the domain name resolution
-     | TraceLocalRootReconfigured [(Int, Map RelayAccessPoint PeerAdvertise)] -- ^ Old value
-                                  [(Int, Map RelayAccessPoint PeerAdvertise)] -- ^ New value
+     | TraceLocalRootReconfigured [( HotValency
+                                   , WarmValency
+                                   , Map RelayAccessPoint PeerAdvertise)] -- ^ Old value
+                                  [( HotValency
+                                   , WarmValency
+                                   , Map RelayAccessPoint PeerAdvertise)] -- ^ New value
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:92]: I wonder how useful it is to trace the old value. I've been reading the logs recently and I was mostly ignoring the previous value.  It's in the logs already but further back in the log.

**Review by coot — CHANGES_REQUESTED** [review]

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs (hunk 13 above)
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs:96]: That's right! Thank you for noticing this!

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs (hunk 15 above)
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:92]: I like it if one is parsing the logs and wants to run some diff, it should be easier this way. Is not like logs should have *less* information if possible

**Review by coot — APPROVED** [review]
LGTM

**Pull request merged**